src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,606 @@
+ *
+ * In the future, the user will be able to override these defaults
+ * to omit unused components (e.g. provide only the term to be integrated against test function gradients)
+ * @tparam element_type The type of the element (used to determine the family)
+ * @tparam T The type of the output from the user-provided q-function
+ * @pre T must be a pair type for H1, H(curl) and H(div) family elements
+ * @param[in] f The value component output of the user's quadrature function (as opposed to the value/derivative pair)
+ * @param[in] xi The position of the quadrature point in reference space
+ * @param[in] J The Jacobian of the element transformation at the quadrature point
+ */
+template <typename element_type, typename T, int dim>
+auto Postprocess(T f, const tensor<double, dim> xi, const tensor<double, dim, dim> J)
+{
+  // TODO: Helpful static_assert about f being tuple or tuple-like for H1, hcurl, hdiv
+  if constexpr (element_type::family == Family::H1) {
+    auto W     = element_type::shape_functions(xi);
+    auto dW_dx = dot(element_type::shape_function_gradients(xi), inv(J));
+    return outer(W, std::get<0>(f)) + dot(dW_dx, std::get<1>(f));
+  }
+
+  if constexpr (element_type::family == Family::HCURL) {
+    auto W      = dot(element_type::shape_functions(xi), inv(J));
+    auto curl_W = element_type::shape_function_curl(xi) / det(J);
+    if constexpr (dim == 3) {
+      curl_W = dot(curl_W, transpose(J));
+    }
+    return (W * std::get<0>(f) + curl_W * std::get<1>(f));
+  }
+}
+
+/**
+ * @overload
+ * @note This specialization of detail::Postprocess is called when doing integrals
+ * where the spatial dimension is different from the dimension of the element geometry
+ * (i.e. surface integrals in 3D space, line integrals in 2D space, etc)
+ *
+ * In this case, q-function outputs are only integrated against test space shape functions
+ * QUESTION: Should test function gradients be supported here or not?
+ */
+template <typename element_type, typename T, int geometry_dim, int spatial_dim>
+auto Postprocess(T f, const tensor<double, geometry_dim> xi,
+                 [[maybe_unused]] const tensor<double, spatial_dim, geometry_dim> J)
+{
+  if constexpr (element_type::family == Family::H1) {
+    return outer(element_type::shape_functions(xi), f);
+  }
+
+  if constexpr (element_type::family == Family::HCURL) {
+    return outer(element_type::shape_functions(xi), dot(inv(J), f));
+  }
+}
+
+/**
+ * @brief Takes in a Jacobian matrix and computes the
+ * associated length / area / volume ratio of the transformation
+ *
+ * In general, this is given by sqrt(det(J^T * J)), but for the case
+ * where J is square, this is equivalent to just det(J)
+ *
+ * @param[in] A The Jacobian to compute the ratio on
+ */
+
+template <int n>
+auto Measure(const tensor<double, n, n>& A)
+{
+  return det(A);
+}
+
+template <int m, int n>
+auto Measure(const tensor<double, m, n>& A)
+{
+  return ::sqrt(det(transpose(A) * A));
```

Comment:
Minor - I think I missed this one when I went through and added this little trick for making sure calls to `sqrt` resolve correctly:

```suggestion
  using std::sqrt;
  return sqrt(det(transpose(A) * A));
```

Reply:
Is this any different than `return std::sqrt(det(transpose(A) * A));`?

Reply:
I would prefer `return std::sqrt(det(transpose(A) * A));` if they are equivalent.

Reply:
I used this trick elsewhere (in `dual.hpp` I think) for functions that have `double` overloads in `std::` but overloads for other types in our code.  I think in this particular function we don't need the trick but for consistency it might be worth keeping - if we ever need to take the `sqrt` of a `dual`, for example, we'll need this for the compiler to be able to resolve both `serac::sqrt(dual)` and `std::sqrt(double)` from the same piece of code.  I know there's an article on this somewhere, I'll try to find it.